lib/src/main/java/io/ably/lib/transport/ConnectionManager.java

```diff
@@ -483,11 +491,22 @@ private synchronized ErrorInfo waitForChange() {
          * ConnectionStateListener interface
          */
         @Override
-        public void onConnectionStateChanged(ConnectionStateChange state) {
-            synchronized(this) {
-                change = state;
-                notify();
+        public synchronized void onConnectionStateChanged(ConnectionStateChange state) {
+            change = state;
+            notify();
+        }
+
+        /**
+         * Remove this ConnectionWaiter as a connection listener.
+         */
+        private void close() {
+            // This method is explicitly not synchronized. There may be a case for this in the
+            // future, however its addition is designed to be lightweight with minimal impact.
+            if (closed) {
```

Comment:
I think adding `synchronized` would be safe, but it this case it happens not to be necessary; if in the race the `.off()` is called twice, the second call would have no effect. Also, the calling code cannot end up calling `close()` more than once.